src/components/Swap/SwapActionButton/Permit2Button.tsx

```diff
@@ -0,0 +1,97 @@
+import { t, Trans } from '@lingui/macro'
+import { PERMIT2_ADDRESS } from '@uniswap/permit2-sdk'
+import ActionButton from 'components/ActionButton'
+import EtherscanLink from 'components/EtherscanLink'
+import { usePendingApproval } from 'hooks/transactions'
+import { PermitState } from 'hooks/usePermit2'
+import { Spinner } from 'icons'
+import { useCallback, useEffect, useMemo, useState } from 'react'
+import { InterfaceTrade } from 'state/routing/types'
+import { ApprovalTransactionInfo } from 'state/transactions'
+import { Colors } from 'theme'
+import { ExplorerDataType } from 'utils/getExplorerLink'
+
+/**
+ * An approving PermitButton.
+ * Should only be rendered if a valid trade exists that is not yet permitted.
+ */
+export default function PermitButton({
+  color,
+  trade,
+  state,
+  callback,
+  onSubmit,
+}: {
+  color: keyof Colors
+  trade?: InterfaceTrade
+  state: PermitState
+  callback?: () => Promise<ApprovalTransactionInfo | void>
+  onSubmit: (submit?: () => Promise<ApprovalTransactionInfo | void>) => Promise<void>
+}) {
+  const currency = trade?.inputAmount?.currency
+  const [isPending, setIsPending] = useState(false)
+  const [isFailed, setIsFailed] = useState(false)
+  useEffect(() => {
+    // Reset pending/failed state if currency changes.
+    setIsPending(false)
+    setIsFailed(false)
+  }, [currency])
+
+  const onClick = useCallback(async () => {
+    setIsPending(true)
+    try {
+      await onSubmit(callback)
+      setIsFailed(false)
+    } catch (e) {
+      console.error(e)
+      setIsFailed(true)
+    } finally {
+      setIsPending(false)
+    }
+  }, [callback, onSubmit])
+
+  const pendingApproval = usePendingApproval(currency?.isToken ? currency : undefined, PERMIT2_ADDRESS)
+
+  const actionProps = useMemo(() => {
+    switch (state) {
```

Comment:
is this to force that all states are checked?

Reply:
yes

Reply:
removed